Add listing of followed hashtags

connorshea opened the pull request:
Resolves #20763 and #20132. Follows up on the addition of hashtag following in #18862.

The ability to follow hashtags was added in v4.0, but there was no addition of functionality for listing the hashtags that the current user is following. This resolves that problem, so users can view and manage their followed hashtags more easily.

This PR doesn't let you follow/unfollow from the list view because I had enough trouble just getting this to work in the first place, and it didn't really seem necessary for an MVP of the list view. It should be relatively easy for someone to go in and add the follow/unfollow buttons to the Hashtag component, or maybe create a distinct component for that purpose.

In the sidebar dropdown:

<img width="345" alt="Screen Shot 2022-11-27 at 8 59 26 PM" src="https://user-images.githubusercontent.com/2977353/204191220-b41efd2f-0b06-4e57-9098-365414d639bd.png">

In the Profile page dropdown:

<img width="219" alt="Screen Shot 2022-11-27 at 8 59 53 PM" src="https://user-images.githubusercontent.com/2977353/204191249-2a17c1da-c057-4add-a69c-e8f7b1bb57cc.png">

"Followed hashtags" list view:

<img width="1224" alt="Screen Shot 2022-11-27 at 8 39 32 PM" src="https://user-images.githubusercontent.com/2977353/204190846-72c350eb-bf88-43e9-a0c6-5e57612bb7a4.png">

Empty state for the "Followed hashtags" list view:

<img width="1217" alt="Screen Shot 2022-11-27 at 9 00 28 PM" src="https://user-images.githubusercontent.com/2977353/204191295-e8bc9d83-2871-43aa-a1b3-a73e9b1696f7.png">

connorshea committed:
Add followed_tags route.

This at least gets us to the point where the page can actually be
rendered, although it doesn't display any hashtags (yet?).

Attempting to implement #20763.

connorshea committed:
Fix minor issues.

connorshea committed:
I've got the followed tags data partially working

But the Hashtag component errors for some reason. Something about the
value of the history attribute being invalid.

connorshea committed:
Fix a mistake in the code

connorshea committed:
Minor change.

connorshea committed:
Get the followed hashtags list fully working.

Still need to add the Follow/Unfollow buttons, though.

connorshea committed:
Resolve JS linter issues.

ClearlyClaire commented on app/javascript/mastodon/features/account/components/header.js:
```diff
@@ -240,6 +241,7 @@ class Header extends ImmutablePureComponent {
       menu.push({ text: intl.formatMessage(messages.follow_requests), to: '/follow_requests' });
       menu.push({ text: intl.formatMessage(messages.favourites), to: '/favourites' });
       menu.push({ text: intl.formatMessage(messages.lists), to: '/lists' });
+      menu.push({ text: intl.formatMessage(messages.followed_tags), to: '/followed_tags' });
```
Not really an issue with your PR, but I'm noticing just now that there are lots of inconsistencies between this dropdown and the composer one. I'm not sure it makes sense to replicate all of this in both and risk having them out of sync…

connorshea replied:
Should we open an issue to extract this into its own component to reuse in both places?

ClearlyClaire reviewed (changes requested):
Thank for your contribution!

It's a good start but it looks like it will only handle one page (100 tags) of followed hashtags, after which it won't show that there are more, nor allow to fetch more.

I think it should handle pagination before being merged.

As for the ability to unfollow a tag from the list itself, it would be a good improvement, but I do not think it is required for this PR, and may require some additional design work.

connorshea committed:
Add pagination logic to followed tags list view.

However, it currently loads further pages immediately on page load, so
that's not ideal. Need to figure that one out.

connorshea committed:
Appease the linter.

connorshea commented:
> It's a good start but it looks like it will only handle one page (100 tags) of followed hashtags, after which it won't show that there are more, nor allow to fetch more.
> 
> I think it should handle pagination before being merged.

I've added pagination now, however - for some reason - it appears to load every page pretty much immediately. Without the user even scrolling to the bottom of the container (or clicking a "Load more" button, though I'm not sure how I'd add that to the ScrollableList component anyway). Any ideas on why that'd happen?

And it's definitely loading multiple pages, because I'm following 318 hashtags locally.

ClearlyClaire commented on app/javascript/mastodon/features/followed_tags/index.js:
```diff
@@ -0,0 +1,96 @@
+    const emptyMessage = <FormattedMessage id='empty_column.followed_tags' defaultMessage='You have not followed any hashtags yet. When you do, they will show up here.' />;
+
+    return (
+      <Column>
```
It's lacking the `bindToDocument` property.

```suggestion
      <Column bindToDocument={!multiColumn}>
```

ClearlyClaire commented on app/javascript/mastodon/features/followed_tags/index.js:
```diff
@@ -0,0 +1,96 @@
+        <ColumnHeader
+          icon='hashtag'
+          title={intl.formatMessage(messages.heading)}
+          showBackButton
```
It's missing the `multiColumn` prop.

```suggestion
          showBackButton
          multiColumn={multiColumn}
```

ClearlyClaire commented on app/javascript/mastodon/features/followed_tags/index.js:
```diff
@@ -0,0 +1,96 @@
+        <ScrollableList
+          scrollKey='followed_tags'
+          emptyMessage={emptyMessage}
+          hasMore={hasMore}
```
It's missing the `isLoading` prop, which plays a role in preventing fetching multiple pages at once, although it does not seem to be the only issue here.

```suggestion
          hasMore={hasMore}
          isLoading={isLoading}
```

ClearlyClaire commented on app/javascript/mastodon/features/followed_tags/index.js:
```diff
@@ -0,0 +1,96 @@
+          scrollKey='followed_tags'
+          emptyMessage={emptyMessage}
+          hasMore={hasMore}
+          onLoadMore={this.handleLoadMore}
```
It's also missing the `bindToDocument` prop:

```suggestion
          onLoadMore={this.handleLoadMore}
          bindToDocument={!multiColumn}
```

ClearlyClaire reviewed (changes requested):
Thank you. I can reproduce the issue and I'm investigating. One thing that you are missing is the `isLoading` prop, but there seem to be more to it than just that.

ClearlyClaire commented on app/javascript/mastodon/reducers/followed_tags.js:
```diff
@@ -0,0 +1,44 @@
+export default function followed_tags(state = initialState, action) {
+  switch(action.type) {
+  case FOLLOWED_HASHTAGS_FETCH_REQUEST:
+    return state.setIn(['isLoading'], true);
+  case FOLLOWED_HASHTAGS_FETCH_SUCCESS:
+    return state.withMutations(map => {
+      map.setIn(['items'], fromJS(action.followed_tags));
+      map.setIn(['isLoading'], false);
+      map.setIn(['next'], action.next);
+    });
+  case FOLLOWED_HASHTAGS_FETCH_FAIL:
+    return state.setIn(['isLoading'], false);
+  case FOLLOWED_HASHTAGS_EXPAND_REQUEST:
+    return state.setIn(['isLoading'], true);
+  case FOLLOWED_HASHTAGS_EXPAND_SUCCESS:
+    return state.withMutations(map => {
+      map.updateIn(['items'], set => set.concat(fromJS(action.followed_tags)));
+      map.setIn(['isLoading'], false);
+      map.setIn(['next'], action.next);
+    });
+  case FOLLOWED_HASHTAGS_EXPAND_FAIL:
+    return state.setIn(['isLoading'], false);
```
Also, not a big deal, but you could use `set(foo, bar)` instead of `setIn([foo], bar)` for all of these, and similarly for `updateIn`.

ClearlyClaire commented on app/javascript/mastodon/features/followed_tags/index.js:
```diff
@@ -0,0 +1,96 @@
+  render () {
+    const { intl, hashtags, isLoading, hasMore } = this.props;
+
+    if (isLoading) {
+      return (
+        <Column>
+          <LoadingIndicator />
+        </Column>
+      );
+    }
```
This is unnecessary and I think it's part of the problem where all items get fetched immediately: fetching more temporarily empties the column.

ClearlyClaire commented on app/javascript/mastodon/features/followed_tags/index.js:
```diff
@@ -0,0 +1,96 @@
+    hasMore: PropTypes.bool,
+  };
+
+  constructor(props) {
+    super(props);
+
+    this.props.dispatch(fetchFollowedHashtags());
+  };
```
I would move that to `componentDidMount`, or maybe `componentWillMount`. But others may disagree with me on this.

connorshea replied:
componentWillMount is deprecated, which is why I went with this pattern, but I'll do componentDidMount

ClearlyClaire commented on app/javascript/mastodon/reducers/followed_tags.js:
```diff
@@ -0,0 +1,44 @@
+import { Map as ImmutableMap, List as ImmutableList, fromJS } from 'immutable';
+
+const initialState = ImmutableMap({
+  followed_tags: ImmutableMap({
+    items: ImmutableList(),
+    isLoading: false,
+    next: null,
+  }),
```
This is incorrectly nested:
```suggestion
  items: ImmutableList(),
  isLoading: false,
  next: null,
```

connorshea committed:
Apply suggestions from code review

Co-authored-by: Claire <claire.github-309c@sitedethib.com>

connorshea committed:
Fixes and resolve some other feedback.

connorshea committed:
Use set/update instead of setIn/updateIn.

connorshea commented:
Had to force-push because I didn't really want to resolve all the merge conflicts from the locale file changes. Instead I just dropped the locales commit, merged develop in, and then regenerated the locale files.

connorshea commented:
@ClearlyClaire Thanks for the help! It's working now, and I think I've resolved all your feedback :)

CEbbinghaus reviewed (approved):
Looks good, No obvious problems

connorshea commented:
Rebased to fix the merge conflicts, I'll leave the updated locale files out for now since I don't think it's worth the trouble with the potential for further conflicts.

@ClearlyClaire this is ready for review, when you have time :)

limdingwen commented:
Can't wait for this! Thank you for your work, @connorshea and @ClearlyClaire.

adammpkins commented:
Any word on whether this can or will be merged?
